mysql-test/include/have_rocksdb_as_default.inc

```
@@ -0,0 +1,12 @@
+--source include/have_rocksdb.inc
```

Maybe push --default-storage-engine=rocksdb to suite.opt too? and then have_rocksdb_as_default.inc becomes redundant? But it needs to be checked what --d-s-e does for an unavailable SE, so that we don't silently test InnoDB, and I'm not sure what to do about `set global rocksdb_force_flush_memtable_now = true;`

---

Hmm, yes. That's actually how I started this but since I was missing the --plugin-load on the mtr call it was skipped/ignored during the collect as we just got through figuring. So assuming it works correctly, it would eliminate the need for the extra mtr options --d-s-e and --d-t-s-e.

From my earlier testing, the proposed have_rocksdb.inc would 'pass' and run the test with a different default engine. This is specifically why I introtuced the have_rocksdb_as_default.inc to ensure that the tests are actually being run as intended. I think I would like to keep that as is since I imagine we will have a transition period where some tests will be written correctly to explicitly use a storage engine, only needing have_rocksdb.inc, and these current tests that depend on rocksdb being default going into the test.

IMHO the net change here should be just to add --d-s-e and --d-t-s-e to the suite.opt and keep the use of have_rocksdb.inc and have_rocksdb_as_default.inc as is. With that I believe we will be guaranteed that if rocksdb failed to load for any reasons, any new explicit test will fail on trying to do something like CREATE TABLE ... ENGINE=ROCKSDB and the current assuming tests will fail/skip on have_rocksdb_as_default.inc.

Maybe, if we wanted to assert it a bit more we could change have_rocksdb_as_default.inc --error rather than --skip on `SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.ENGINES WHERE engine = 'rocksdb' AND support IN ('DEFAULT')`.

As far as `set global rocksdb_force_flush_memtable_now = true`, IMHO it is another bad move by the test writers and should be placed within specific tests that rely on deterministic behavior of table statistics, not in a global catch-all. I will create a new JIRA task for this and see if I can figure out how many tests would need changing and possibly suggest it upstream. For now, I will just add JIRA tags to this with the new issue id until I can figure out how to sort it through the individual test cases.

---

So testing this out and it doesn't work with the suite.opt as expected. suite.opt is `$ROCKSDB_OPT $ROCKSDB_LOAD_ADD --default-storage-engine=rocksdb --default-tmp-storage-engine=rocksdb`

and mtr incantation is `./mtr  --force --retry=0 --retry-failure=0 --max-test-fail=0 --no-warnings --testcase-timeout=600 --mysqld=--plugin-load="rocksdb=ha_rocksdb.so;rocksdb_cfstats=ha_rocksdb.so;rocksdb_dbstats=ha_rocksdb.so;rocksdb_perf_context=ha_rocksdb.so;rocksdb_perf_context_global=ha_rocksdb.so;rocksdb_cf_options=ha_rocksdb.so;rocksdb_global_info=ha_rocksdb.so;rocksdb_ddl=ha_rocksdb.so;rocksdb_index_file_map=ha_rocksdb.so" --suite=rocksdb`

Adding the `--mysqld=--default-storage-engine=rocksdb --mysqld=--default-tmp-storage-engine=rocksdb` back in works. The logs indicate it is the have_rocksdb_as_default.inc that is causing the skip. So somehow the --d-s-e and --d-t-s-e suite options are not getting passed into the server instance.

From a little poking and debugging, it looks like in  $default_storage_engine is set if specified on the command line, but it is not set if it is in suite.opt, this causes the logic in collect_one_test_case to take a different path and I believe the end result is that MyISAM becomes the default engine. So it seems that options in the suite.opts and test.opts are not run through the collect_options function which sets the global option variables. That makes sense because we wouldn't want a single test or suite option to change the global for other suites. So to fix this I think there would need to be a test by test check in the options for things like --d-s-e and perhaps others.

Honestly, at this point I'm not sure the depth of changes needed to fix this are worth the effort and risk of breaking mtr. Can you confirm my research here and advise if we should just leave as is and require --d-s-e and --d-t-s-e along with the --mysqld=--plugin-load="..." incantation in order to run rocksdb tests for now and perhaps revisit this at some later time if it becomes more critical? Perhaps we can convince upstream to fix their test cases to not pre-require rocksdb as the d-s-e which would solve it all for us anyway.

---

Yes, everything you suggest sounds good. 

BTW, if you are patching the testcases for s/have_rocksdb/have_rocksdb_as_default/g, then patching for SET default_storage_engine=... is similar amount of patching, isn't it :) I'm not saying to make this change, it's just something that could be considered in the future

---

Ah, have_rocksdb_as_default.inc can do SET default_storage_engine = ... itself, can't it?

---

No it cannot, this will "fail" instead of "skip" on -DWITHOUT_ROCKSDB=ON

---

Correct, and in order for it to function properly for multi-session tests, it would have to set global, which then also means we would need a 'clean up' step at the end of each test to reset it back to previous state.

---

Also, changing `--source include/have_rocksdb.inc` to `--source include/have_rocksdb_as_default.inc` was just a simple sed. Adding the whole `get current d-e-s; set d-s-e=rocksdb; <test body>; set d-s-e=old-d-s-e` is quite a bit more interesting to ensure it is correct on all tests.